src/val.rs

```diff
@@ -0,0 +1,128 @@
+use std::rc::Rc;
+
+// Design notes:
+// =============
+//
+// The "list of strings" data structure can be represented in three different reasonable ways:
+//
+// 1. As a vec of strings (currently implemented).
+// This is simple and easy to debug, but makes zero-copy stuff difficult.
+// At least can make copying cheap with an Rc<T>.
+//
+// 2. As a pre-joined string + vec of indices.
+// This allows a `&str` ref to the pre-joined string to be acquired,
+// at the cost of bounds checking when the parts are accessed.
+// Would probably be the most compact and efficient implementation.
+//
+// 3. As a value that contains both a pre-joined variant and list variant.
+// This trades memory for the ability to borrow refs to both variants.
+//
+// The current implementation focuses on simplicity,
+// but tries to preserve flexibility for future optimizations
+// by mostly only dealing with owned data at the API boundaries (no borrowed views).
+
+/// The internal data type of Just, an immutable list of strings.
+///
+/// It can be viewed as a "joined string", or as the individual "parts".
+#[derive(Debug, Clone, Default)]
+pub(crate) struct Val {
+  parts: Rc<[Box<str>]>,
+}
+
+impl Val {
+  /// Create an empty `Val`
+  pub fn new() -> Self {
+    Self::default()
+  }
+
+  /// Construct a `Val` consisting of a single string.
+  pub fn from_str<S: AsRef<str>>(s: S) -> Self {
+    Self {
+      parts: Rc::new([s.as_ref().into()]),
+    }
+  }
+
+  /// Construct a `Val` consisting of multiple parts.
+  pub fn from_parts<I, S>(parts: I) -> Self
+  where
+    I: IntoIterator<Item = S>,
+    S: AsRef<str>,
+  {
+    Self {
+      parts: parts.into_iter().map(|s| s.as_ref().into()).collect(),
+    }
+  }
+
+  /// Convert to a single joined string.
+  pub fn to_joined(&self) -> String {
+    self.parts.join(" ")
+  }
+
+  /// Convert to individual parts.
+  pub fn to_parts(&self) -> Vec<Box<str>> {
```

Comment:
Can this return just `&[String]`, and rely on the caller to convert it to a vec? Avoids allocation if the caller doesn't need to.